File: app/src/main/java/com/qiaobateach/qiaostander/LoadPluginActivity.java
package com.qiaobateach.qiaostander;

import android.app.Activity;
import android.content.Intent;
import android.content.res.Resources;
import android.os.Bundle;
import android.support.annotation.Nullable;

import com.qiao_stander.QiaoInterface;

import java.lang.reflect.Constructor;

public class LoadPluginActivity extends Activity{


    private String className;
    private QiaoInterface mQiaoInterface;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        className = getIntent().getStringExtra("className");

        try {
            Class activity = getClassLoader().loadClass(className);
            Constructor constructor = activity.getConstructor(new Class[]{});
            Object instance = constructor.newInstance(new Object[]{});
            mQiaoInterface = (QiaoInterface)instance;
            mQiaoInterface.attach(this);
            Bundle bundle = new Bundle();
            mQiaoInterface.onCreate(bundle);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        mQiaoInterface.onStart();
    }


    @Override
    protected void onResume() {
        super.onResume();
        mQiaoInterface.onResume();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mQiaoInterface.onDestroy();
    }


    @Override
    protected void onPause() {
        super.onPause();
        mQiaoInterface.onPause();
    }

    @Override
    protected void onStop() {
        super.onStop();
        mQiaoInterface.onStop();
    }

    @Override
    public ClassLoader getClassLoader() {
        return PluginLoadManager.getInstance().getmDexClassLoader();
    }

    @Override
    public Resources getResources() {
        return PluginLoadManager.getInstance().getmResources();
    }


    @Override
    public void startActivity(Intent intent) {
        String className = intent.getStringExtra("className");
        /***
         * 跳转自己，为了重新赋予插件要跳转的activity的生命周期
         */
        Intent i = new Intent(this,LoadPluginActivity.class);
        i.putExtra("className",className);
        super.startActivity(i);
    }
}

File: app/src/main/java/com/qiaobateach/plugin/LoadPluginActivity.kt
package com.qiaobateach.plugin

import android.app.Activity
import android.content.Intent
import android.content.res.Resources
import android.os.Bundle

import com.qiao_stander.QiaoInterface
import dalvik.system.DexClassLoader

import java.lang.reflect.Constructor

class LoadPluginActivity : Activity() {


    private var className: String? = null
    private var mQiaoInterface: QiaoInterface? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        className = intent.getStringExtra("className")

        try {
            val activity = classLoader!!.loadClass(className)
            val constructor = activity.getConstructor(*arrayOf())
            val instance = constructor.newInstance(*arrayOf())
            mQiaoInterface = instance as QiaoInterface
            mQiaoInterface!!.attach(this)
            val bundle = Bundle()
            mQiaoInterface!!.onCreate(bundle)
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onStart() {
        super.onStart()
        mQiaoInterface!!.onStart()
    }


    override fun onResume() {
        super.onResume()
        mQiaoInterface!!.onResume()
    }

    override fun onDestroy() {
        super.onDestroy()
        mQiaoInterface!!.onDestroy()
    }


    override fun onRestart() {
        super.onRestart()

    }


    override fun onPause() {
        super.onPause()
        mQiaoInterface!!.onPause()
    }

    override fun onStop() {
        super.onStop()
        mQiaoInterface!!.onStop()
    }

    override fun getClassLoader(): DexClassLoader? {
        return PluginLoadManager.instance.getmDexClassLoader();
//        return PluginLoadManager.getInstance().getmDexClassLoader()
    }

    override fun getResources(): Resources? {
        return PluginLoadManager.instance.getmResources();
    }


    override fun startActivity(intent: Intent) {
        val className = intent.getStringExtra("className")
        /***
         * 跳转自己，由代理Activity跳转到目标页面，为了重新赋予插件要跳转的activity的生命周期
         */
        val i = Intent(this, LoadPluginActivity::class.java)
        i.putExtra("className", className)
        super.startActivity(i)
    }
}
